app/code/Bulbulatory/Recommendations/Controller/Customer/ConfirmRecommendation.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+namespace Bulbulatory\Recommendations\Controller\Customer;
+
+use Bulbulatory\Recommendations\Helper\Acl;
+use Exception;
+use Magento\Framework\App\Action\Context;
+use Magento\Framework\App\Action\Action;
+use Bulbulatory\Recommendations\Model\RecommendationRepository;
+use Psr\Log\LoggerInterface;
+
+class ConfirmRecommendation extends Action
+{
+    /**
+     * @var RecommendationRepository
+     */
+    private $recommendationsRepository;
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    /**
+     * @var
+     */
+    private $hash;
+    /**
+     * @var Acl
+     */
+    private $aclHelper;
+
+    /**
+     * ConfirmRecommendation constructor.
+     * @param Context $context
+     * @param RecommendationRepository $recommendationRepository
+     * @param LoggerInterface $logger
+     * @param Acl $aclHelper
+     */
+    public function __construct(
+        Context $context,
+        RecommendationRepository $recommendationRepository,
+        LoggerInterface $logger,
+        Acl $aclHelper
+    )
+    {
+        parent::__construct($context);
+
+        $this->recommendationsRepository = $recommendationRepository;
+        $this->logger = $logger;
+        $this->aclHelper = $aclHelper;
+    }
+
+    /**
+     * @return mixed
+     */
+    public function execute()
+    {
+        if(!$this->aclHelper->recommendationModuleAccess())
+        {
+            $this->messageManager->addErrorMessage(__('You have not access for confirm recommendation!'));
+            return $this->resultRedirectFactory->create()->setPath('/');
+        }

```

Comment:
This is not ACL, this is not a situation when someone doesn't have access to some part of the application and someone does. This is a check if the module **is enabled** or not. So the name `aclHelper` and method `recommendationModuleAccess` are misleading a bit. All methods related to fetching configuration value should be put in one file eg. `configHelper` and a more appropriate name for the method would be `isEnabled`, `isRecommendationModuleEnabled`, `isModuleEnabled` or something like that explicitly saying that we check if it's enabled or not.

The purpose of the configuration allowing to enable or disable a module is that the store owner can do it whenever he/she wants. That's why you shouldn't put any error messages if it's not enabled as customers shouldn't know it - just redirect to 404 or homepage (as you did) without displaying any information to the end user.

By the way this check is in other controllers as well so what do you think about creating an abstract class where this check will happen? Then these controller classes can extend it and the logic checking if the module is enabled will be only in the abstract class instead of duplicating it in other classes.

Reply:
@TM18

> This is not ACL, this is not a situation when someone doesn't have access to some part of the application and someone does. This is a check if the module is enabled or not. So the name aclHelper and method recommendationModuleAccess are misleading a bit. All methods related to fetching configuration value should be put in one file eg. configHelper and a more appropriate name for the method would be isEnabled, isRecommendationModuleEnabled, isModuleEnabled or something like that explicitly saying that we check if it's enabled or not.

I removed class **ACl** and replace **recommendationModuleAccess** method to **Config** class and rename this method to **isModuleEnabled**

> The purpose of the configuration allowing to enable or disable a module is that the store owner can do it whenever he/she wants. That's why you shouldn't put any error messages if it's not enabled as customers shouldn't know it - just redirect to 404 or homepage (as you did) without displaying any information to the end user.

I removed error messages for customer

> By the way this check is in other controllers as well so what do you think about creating an abstract class where this check will happen? Then these controller classes can extend it and the logic checking if the module is enabled will be only in the abstract class instead of duplicating it in other classes.

I created RecommendationAbstract class.